p2panda-discovery-next/src/naive.rs

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::collections::{BTreeMap, HashSet};
+use std::hash::Hash as StdHash;
+use std::marker::PhantomData;
+
+use thiserror::Error;
+use tokio::sync::mpsc;
+
+use crate::address_book::{AddressBookStore, NodeInfo};
+use crate::traits::{DiscoveryProtocol, DiscoveryResult, Receiver, Sender, SubscriptionInfo};
+
+pub enum NaiveDiscoveryMessage<T, ID, N>
+where
+    N: NodeInfo<ID>,
+{
+    Topics {
+        topics: HashSet<T>,
+        topic_ids: HashSet<[u8; 32]>,
+    },
+    Nodes {
+        transport_infos: BTreeMap<ID, N::Transports>,
+    },
+}
+
+pub struct NaiveDiscoveryProtocol<S, P, T, ID, N> {
+    store: S,
+    subscription: P,
+    remote_node_id: ID,
+    _marker: PhantomData<(T, N)>,
+}
+
+impl<S, P, T, ID, N> NaiveDiscoveryProtocol<S, P, T, ID, N> {
+    pub fn new(store: S, subscription: P, remote_node_id: ID) -> Self {
+        Self {
+            store,
+            subscription,
+            remote_node_id,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<S, P, T, ID, N> DiscoveryProtocol<T, ID, N> for NaiveDiscoveryProtocol<S, P, T, ID, N>
+where
+    S: AddressBookStore<T, ID, N>,
+    P: SubscriptionInfo<T>,
+    T: Eq + StdHash,
+    ID: Clone + Ord,
+    N: NodeInfo<ID>,
+{
+    type Error = NaiveDiscoveryError<S, P, T, ID, N>;
+
+    type Message = NaiveDiscoveryMessage<T, ID, N>;
+
+    async fn alice(
+        &self,
+        tx: Sender<Self::Message>,
+        mut rx: Receiver<Self::Message>,
+    ) -> Result<DiscoveryResult<T, ID, N>, Self::Error> {
+        // 1. Alice sends Bob all their topics and topic ids.
```

Comment:
The lib level description is really great and I get a clear understanding of the core types and features of this crate. However when I read this line and code here I get a little confused as the `T` topic is now introduced whereas in the doc strings a topic was only ever referred to as " a secret, randomly generated hash, similar to a shared symmetric key". 

From my perspective working in p2panda-sync (where the T topic is used) I'm not sure what the reason for sending it here in discovery is. That is maybe for a wider conversation about how the systems fit together though, which we will have soon.

In any case, it would be good to include some explanation in the docs here of what the T topic is and what peers gain from discovering it here.